File: Python/spacetime/basis.py
from ..datastructures.double_tree_view import DoubleTree
from ..space.basis import HierarchicalBasisFunction
from ..time.orthonormal_basis import OrthonormalBasis
from ..time.three_point_basis import ThreePointWavelet


def generate_x_delta_underscore(x_delta):
    """ Generates X^{underline delta} as in p.27 from followup3.pdf. """
    assert isinstance(x_delta, DoubleTree)
    assert isinstance(x_delta.root.nodes[0].children[0], ThreePointWavelet)
    assert isinstance(x_delta.root.nodes[1].children[0],
                      HierarchicalBasisFunction)

    x_delta_underscore = x_delta.deep_copy()

    dblnodes = x_delta_underscore.bfs()
    time_leaves = []
    space_leaves = []
    for dblnode in dblnodes:
        # First, we ensure that the underlying trees are properly refined.
        if not dblnode.nodes[0].is_full():
            dblnode.nodes[0].refine()
        if not dblnode.nodes[1].node.is_full() or not dblnode.nodes[1].is_full(
        ):
            dblnode.nodes[1].node.refine()
            dblnode.nodes[1].refine(make_conforming=True)

        # The first part of this if-statement fends off the situation where
        # dblnode.children[i] has 0 < n < full elements as a result of adaptive
        # refinement in X_delta.
        if not dblnode.is_full(0):
            # Refine in time-axis...
            time_leaves.append(dblnode)

        if not dblnode.is_full(1):
            # and double-refine in space-axis.
            space_leaves.append(dblnode)

    for dblnode in time_leaves:
        dblnode.refine(i=0, make_conforming=True)

    for dblnode in space_leaves:
        children = dblnode.refine(i=1, make_conforming=True)
        for child in children:
            child.nodes[1].node.refine()
            child.nodes[1].refine(make_conforming=True)
            child.refine(i=1, make_conforming=True)

    dblnodes_underscore = x_delta_underscore.bfs()
    for dblnode in dblnodes:
        dblnode.marked = True

    new_dblnodes = []
    for dblnode in dblnodes_underscore:
        if dblnode.marked:
            dblnode.marked = False
        else:
            new_dblnodes.append(dblnode)

    return x_delta_underscore


def generate_y_delta(x_delta):
    """ Generates Y^\delta from X^\delta as p.6 from followup3.pdf. """

    assert isinstance(x_delta, DoubleTree)
    assert isinstance(x_delta.root.nodes[0].children[0], ThreePointWavelet)
    assert isinstance(x_delta.root.nodes[1].children[0],
                      HierarchicalBasisFunction)

    y_basis_time = OrthonormalBasis()

    # Create an empty double tree.
    y_delta = DoubleTree.from_metaroots(
        (y_basis_time.metaroot_wavelet, x_delta.root.nodes[1]))

    # Plug in the space metaroot axis.
    y_delta.project(1).union(x_delta.project(1))

    # For the time metaroot axis, we first *mark* which wavelets in time
    # should be tensored with which *trees* in space.
    for x_labda_0 in x_delta.project(0).bfs():
        for elem in x_labda_0.node.support:
            for mu in elem._refine_psi_orthonormal():
                if not mu.marked: mu.marked = []
                mu.marked.append(x_labda_0.frozen_other_axis())

    # First, we use this data to create the time metaroot axis.
    y_delta.project(0)._deep_refine(call_filter=lambda mu: mu.marked)

    # Then, we union the output wavelets with the right space trees.
    for y_labda_0 in y_delta.project(0).bfs():
        assert y_labda_0.node.marked
        for space_tree in y_labda_0.node.marked:
            y_labda_0.frozen_other_axis().union(space_tree)

        # Remove the marks.
        y_labda_0.node.marked = False

    y_delta.compute_fibers()

    return y_delta

File: C++/spacetime/basis.cpp
#include "basis.hpp"

#include <vector>

namespace spacetime {
using datastructures::DoubleTreeVector;
using datastructures::DoubleTreeView;
using space::HierarchicalBasisFn;
using Time::OrthonormalWaveletFn;
using Time::ThreePointWaveletFn;

template <template <typename, typename> class DblTreeIn,
          template <typename, typename> class DblTreeOut>
DblTreeOut<OrthonormalWaveletFn, HierarchicalBasisFn> GenerateYDelta(
    const DblTreeIn<ThreePointWaveletFn, HierarchicalBasisFn> &X_delta) {
  using SpaceTreeVector =
      std::vector<typename DblTreeIn<ThreePointWaveletFn,
                                     HierarchicalBasisFn>::FrozenDN1Type *>;
  // Figure out the orthonormal metaroot.
  OrthonormalWaveletFn *ortho_meta_root = std::get<0>(X_delta.root()->nodes())
                                              ->children()
                                              .at(0)
                                              ->support()
                                              .at(0)
                                              ->RefinePsiOrthonormal()[0]
                                              ->parents()
                                              .at(0);
  assert(ortho_meta_root->is_metaroot());

  auto Y_delta = DblTreeOut<OrthonormalWaveletFn, HierarchicalBasisFn>(
      ortho_meta_root, std::get<1>(X_delta.root()->nodes()));

  Y_delta.Project_1()->Union(X_delta.Project_1());

  for (auto &x_labda_0 : X_delta.Project_0()->Bfs())
    for (auto &elem : x_labda_0->node()->support())
      for (auto &mu : elem->RefinePsiOrthonormal()) {
        if (!mu->has_data()) mu->set_data(new SpaceTreeVector());
        mu->template data<SpaceTreeVector>()->push_back(
            x_labda_0->FrozenOtherAxis());
      }

  Y_delta.Project_0()->DeepRefine([](auto node) { return node->has_data(); });

  for (auto &y_labda_0 : Y_delta.Project_0()->Bfs()) {
    assert(y_labda_0->node()->has_data());
    auto data = y_labda_0->node()->template data<SpaceTreeVector>();
    for (const auto &space_tree : *data)
      y_labda_0->FrozenOtherAxis()->Union(space_tree);
    y_labda_0->node()->reset_data();
    delete data;
  }

  return Y_delta;
}

DoubleTreeVector<ThreePointWaveletFn, HierarchicalBasisFn>
GenerateXDeltaUnderscore(
    const DoubleTreeVector<ThreePointWaveletFn, HierarchicalBasisFn> &X_delta,
    size_t num_repeats) {
  assert(num_repeats > 0);
  if (num_repeats > 1)
    return GenerateXDeltaUnderscore(GenerateXDeltaUnderscore(X_delta),
                                    num_repeats - 1);
  DoubleTreeVector<ThreePointWaveletFn, HierarchicalBasisFn>
      X_delta_underscore = X_delta.DeepCopy();

  std::vector<typename decltype(X_delta_underscore)::DNType *> time_leaves,
      space_leaves;
  for (auto &dblnode : X_delta_underscore.container()) {
    auto [time_node, space_node] = dblnode.nodes();
    if (!time_node->is_full()) time_node->Refine();
    if (!space_node->is_full())  // This is DIFFERENT from python.
      space_node->Refine();

    if (!dblnode.template is_full<0>()) time_leaves.push_back(&dblnode);
    if (!dblnode.template is_full<1>()) space_leaves.push_back(&dblnode);
  }

  for (auto dblnode : time_leaves)
    dblnode->template Refine<0>(datastructures::func_true,
                                /*make_conforming*/ true);

  for (auto dblnode : space_leaves) {
    dblnode->template Refine<1>(datastructures::func_true,
                                /*make_conforming*/ true);
    for (auto child : dblnode->children(1)) {
      child->node_1()->Refine();
      child->template Refine<1>(datastructures::func_true,
                                /*make_conforming*/ true);
    }
  }

  return X_delta_underscore;
}

// Template specializations for GenerateYDelta.
template DoubleTreeView<OrthonormalWaveletFn, HierarchicalBasisFn>
GenerateYDelta<DoubleTreeView, DoubleTreeView>(
    const DoubleTreeView<ThreePointWaveletFn, HierarchicalBasisFn> &X_delta);
template DoubleTreeVector<OrthonormalWaveletFn, HierarchicalBasisFn>
GenerateYDelta<DoubleTreeVector, DoubleTreeVector>(
    const DoubleTreeVector<ThreePointWaveletFn, HierarchicalBasisFn> &X_delta);

};  // namespace spacetime
